### DISCUSSION TITLE
How to mock dynamic paths with vi.mock

### DISCUSSION THREAD

**Discussion by rsaiprasad** [vitest-dev/vitest#4548] [Q&A]
I have a library that has multiple components exported which are to be imported by individual paths. How do I mock in this case. 
Ex:
```
const paths = ['component1', 'component2'... hundreds];
paths.forEach((component) => {
 vi.mock(`@library/components/${component}` , () => {
  default: vi.fn(),
}
});
```
Does not work. 
Is it possible to do something like this? looks not possible for me as vitest statically analyses files for vi.mock. Any thoughts / help is appreciated.

**Comment by sheremet-va** [comment]
Use vi.doMock

**Reply by heyakyra** [reply]
Use it how? It has the same type signature as `vi.mock` which doesn't look like it supports a way to pass in a regex to pattern match against, for example to be able to transport all imports of a certain file extension. 

In the case someone is using [Fluent](https://github.com/heyakyra/fluent-companies?tab=readme-ov-file#fluent-companies), translation files, they will see errors like: `Error: Failed to parse source for import analysis because the content contains invalid JS syntax. You may need to install appropriate plugins to handle the .ftl file format, or if it's an asset, add "**/*.ftl" to ``assetsInclude`` in your configuration.`

and a proper file transformer is needed for this file type which was possible in `jest` using a config line like the last one in this block:

```typescript
  transform: {
    "^.+\\.vue$": "@vue/vue3-jest",
    "^.+\\.(ts|tsx|js)$": "ts-jest",
    "^.+\\.mjs$": "babel-jest",
    "^.+\\.ftl$": "<rootDir>/__mocks__/ftlTransformer.js",
```

combined with a [global auto-mock](https://vitest.dev/guide/migration.html#auto-mocking-behaviour), `__mocks__/ftlTransformer.js` like so:

```typescript
import fs from "fs";

export default {
  process(_sourceText, sourcePath) {
    const content = fs.readFileSync(sourcePath, "utf8");
    return {
      code: `module.exports = ${JSON.stringify(content)};`,
    };
  },
};
```

**Reply by sheremet-va** [reply]
> Use it how?

See the code in question:

```diff
const paths = ['component1', 'component2'... hundreds];
paths.forEach((component) => {
- vi.mock(`@library/components/${component}` , () => {
+ vi.doMock(`@library/components/${component}` , () => {
  default: vi.fn(),
}
});
```

> It has the same type signature as vi.mock which doesn't look like it supports a way to pass in a regex to pattern match against, for example to be able to transport all imports of a certain file extension.

If you need to [alias](https://vitest.dev/guide/mocking.html#virtual-modules) all modules, you can use `alias`:

```ts
import { defineConfig } from 'vitest/config'
import { resolve } from 'node:path'
export default defineConfig({
  test: {
    alias: [
      { find: /\.ftl$/, replacement: resolve('./mocks/ftl.js') },
    ],
  },
})
```

If you need to know the id/code of the module, you can create a Vite plugin:

```ts
import { defineConfig } from 'vitest/config'
export default defineConfig({
  plugins: [
    {
      name: 'replace-ftl',
      transform(code, id) {
        if (id.match(/\.ftl$/)) {
          return `export default ${JSON.stringify(code)}`
        }
      },
    },
  ],
})
```
